app/Command/Graph.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections #-}
+
+module Command.Graph (command) where
+
+import           Control.Applicative (many)
+import           Control.Monad (unless, when)
+import qualified Data.Aeson as Json
+import           Data.Bool (bool)
+import qualified Data.ByteString.Lazy as LB
+import qualified Data.ByteString.Lazy.UTF8 as LBU8
+import qualified Language.PureScript as P
+import           Language.PureScript.Errors.JSON
+import qualified Options.Applicative as Opts
+import qualified System.Console.ANSI as ANSI
+import           System.Exit (exitSuccess, exitFailure)
+import           System.Directory (getCurrentDirectory)
+import           System.FilePath.Glob (glob)
+import           System.IO (hPutStr, hPutStrLn, stderr)
+
+data GraphOptions = GraphOptions
+  { graphInput      :: [FilePath]
+  , graphJSONErrors :: Bool
+  }
+
+graph :: GraphOptions -> IO ()
+graph GraphOptions{..} = do
+  input <- globWarningOnMisses (unless graphJSONErrors . warnFileTypeNotFound) graphInput
+  when (null input && not graphJSONErrors) $ do
+    hPutStr stderr $ unlines
+      [ "purs graph: No input files."
+      , "Usage: For basic information, try the `--help' option."
+      ]
+    exitFailure
+
+  (makeWarnings, makeResult) <- P.graph input
+
+  printWarningsAndErrors True graphJSONErrors makeWarnings makeResult
+
+  case makeResult of
+    Left _ -> exitSuccess
+    Right result -> LB.putStr result
+
+  where
+  warnFileTypeNotFound :: String -> IO ()
+  warnFileTypeNotFound =
+    hPutStrLn stderr . ("purs graph: No files found using pattern: " <>)
+
+
+command :: Opts.Parser (IO ())
+command = graph <$> (Opts.helper <*> graphOptions)
+  where
+  graphOptions :: Opts.Parser GraphOptions
+  graphOptions =
+    GraphOptions <$> many inputFile
+                 <*> jsonErrors
+
+  inputFile :: Opts.Parser FilePath
+  inputFile =
+    Opts.strArgument $
+      Opts.metavar "FILE" <>
+      Opts.help "The input .purs file(s)."
+
+  jsonErrors :: Opts.Parser Bool
+  jsonErrors =
+    Opts.switch $
+      Opts.long "json-errors" <>
+      Opts.help "Print errors to stderr as JSON"
+
+
+-- | Arguments: verbose, use JSON, warnings, errors
+printWarningsAndErrors :: Bool -> Bool -> P.MultipleErrors -> Either P.MultipleErrors a -> IO ()
+printWarningsAndErrors verbose False warnings errors = do
```

Comment:
I think `Either P.MultipleErrors a` might actually make sense here because that's how the compiler represents results everywhere else: `Left` if there are errors, otherwise a `Right` compilation result. Also, while `P.MultipleErrors` does have a `Monoid` instance i.e. it can be empty, we generally treat the existence of one as "errors have occurred" rather than peeking inside to see if it is empty, so I think it would probably be preferable to stick with that scheme here.

Reply:
In that case I think the signature should change to
`printWarningsAndErrors :: Bool -> P.MultipleErrors -> Either P.MultipleErrors a -> IO a` and `exitFailure` should be called in the case of an error.

Reply:
Fixed :+1: